formats/json-okio/commonMain/src/kotlinx/serialization/json/okio/OkioStreams.kt

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2017-2022 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.json.okio
+
+import kotlinx.serialization.*
+import kotlinx.serialization.json.DecodeSequenceMode
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.internal.*
+import kotlinx.serialization.json.okio.internal.JsonToOkioStreamWriter
+import kotlinx.serialization.json.internal.decodeToSequenceByReader
+import kotlinx.serialization.json.okio.internal.OkioSerialReader
+import okio.*
+
+/**
+ * Serializes the [value] with [serializer] into a [target] using JSON format and UTF-8 encoding.
+ *
+ * If [target] is not a [BufferedSink] then called [Sink.buffer] for it to create buffered wrapper.
```

Comment:
Sorry I'm late to the party! Why not making the "buffer" requirement in the API itself? 

The current API fails if someone expects to be able to resume reading the stream after decoding an element.

See https://github.com/Kotlin/kotlinx.serialization/issues/1789#issuecomment-988943550 for a related discussion.

Reply:
For encoding this isn't a problem. You are always free to call `buffer()` and then `emit()` as implementation details.

The problem you are referring to, however, is for decoding. Decoding absolutely should take a `BufferedSource` and not a `Source`. This is what Moshi and Wire do as well.

https://github.com/square/moshi/blob/moshi-parent-1.13.0/moshi/src/main/java/com/squareup/moshi/JsonAdapter.java#L57

https://github.com/square/wire/blob/4.4.0/wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoAdapter.kt#L145

Reply:
Right! Sorry for the confusion, I never realised how asymmetric this was 👍 . Maybe it's worth removing the "BufferSink" mention from the kdoc then? Since it's an implementation detail, the caller doesn't really need to know about it?

And in all cases, make the `BufferedSource` mandatory for `decodeFromSource`:

```kotlin
public fun <T> Json.decodeFromSource(
    deserializer: DeserializationStrategy<T>,
    source: BufferedSource
): T
```

Reply:
Agree on both points!

Reply:
Thanks for your suggestions! Looks like we probably need to change existing functions working with Java streams applying the same logic.

However, this may lead to some asymmetry in the API. What do you think would be better — change both `encodeToSink/decodeFromSource` so they accept buffered versions or leave `encodeToSink` intact?

Reply:
>  change both encodeToSink/decodeFromSource so they accept buffered versions or leave encodeToSink intact?

No strong opinion from me. Both [Moshi](https://github.com/square/moshi/blob/moshi-parent-1.13.0/moshi/src/main/java/com/squareup/moshi/JsonAdapter.java#L80) and [Wire](https://github.com/square/moshi/blob/moshi-parent-1.13.0/moshi/src/main/java/com/squareup/moshi/JsonAdapter.java#L80) use `BufferedSink` so I'd lean towards `BufferedSink` here as well for consistency/aesthetics reasons. 

Since the sink is ending up being buffered in all cases, it's pretty easy to let the caller call `.buffer()` if required. Plus it might even save an `emit()` when writing multiple Json to the same `BufferedSink` (So tiny performance gain maybe...). But if there's a use case for passing a raw `Sink`, that works with me too.